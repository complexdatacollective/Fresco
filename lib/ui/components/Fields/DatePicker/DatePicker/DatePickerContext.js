import React from 'react';

const DatePickerContext = React.createContext({
  onChange: () => {},
  range: null,
  date: {},
  type: null,
});

export default DatePickerContext;
