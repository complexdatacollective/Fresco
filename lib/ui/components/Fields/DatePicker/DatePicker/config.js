/* eslint-disable import/prefer-default-export */

export const DEFAULT_MIN_DATE = { years: 100 }; // DateTime.minus(DEFAULT_MIN_DATE);

export const DEFAULT_TYPE = 'full';

export const DATE_FORMATS = {
  full: 'yyyy-MM-dd',
  month: 'yyyy-MM',
  year: 'yyyy',
};
