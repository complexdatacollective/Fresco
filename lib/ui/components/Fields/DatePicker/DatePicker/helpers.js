/* eslint-disable import/prefer-default-export */

import { DateTime, Info } from 'luxon';
import { isEqual, get } from 'lodash';

export const now = () => DateTime.local();

/**
 * Is date object fully complete?
 */
export const isComplete = (type) => ({ day, month, year }) => {
  switch (type) {
    case 'year':
      return !!year;
    case 'month':
      return !!year && !!month;
    default:
      return !!year && !!month && !!day;
  }
};

/**
 * Is date object empty
 */
export const isEmpty = () => (date) => isEqual(date, { year: null, month: null, day: null });

export const getMonthName = (numericMonth) => get(Info.months(), numericMonth - 1, numericMonth);

export const formatRangeItem = (value, props = {}) => ({
  value,
  label: value,
  ...props,
});
