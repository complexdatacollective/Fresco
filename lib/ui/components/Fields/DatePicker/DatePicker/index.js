export { default as Years } from './Years';
export { default as Months } from './Months';
export { default as Days } from './Days';
export { default as Date } from './Date';
export { default as DatePicker } from './DatePicker';
export { DATE_FORMATS } from './config';
export { now } from './helpers';
