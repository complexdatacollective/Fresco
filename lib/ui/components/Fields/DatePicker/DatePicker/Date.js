import { useContext } from 'react';
import DatePickerContext from './DatePickerContext';
import { isComplete, isEmpty } from './helpers';

const Date = ({ children }) => {
  const {
    onChange, date, range, type,
  } = useContext(DatePickerContext);

  return children({
    onChange,
    date,
    range,
    type,
    isComplete: isComplete(type)(date),
    isEmpty: isEmpty(type)(date),
  });
};

export default Date;
