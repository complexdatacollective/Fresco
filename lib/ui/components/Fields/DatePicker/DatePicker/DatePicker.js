import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { DateTime, Interval } from 'luxon';
import DatePickerContext from './DatePickerContext';
import { DATE_FORMATS, DEFAULT_TYPE, DEFAULT_MIN_DATE } from './config';
import { now, isComplete, isEmpty } from './helpers';

/**
 * Get date object from an ISO string
 */
const getDate = (dateString) => {
  const { year, month, day } = dateString
    ? DateTime.fromISO(dateString).toObject()
    : {
      month: null,
      day: null,
      year: null,
    };
  return { year, month, day };
};

const DatePicker = ({
  children,
  date,
  min,
  max,
  onChange,
  type,
}) => {
  const [pickerState, setPickerState] = useState({
    date: getDate(date),
  });

  // Correctly update component state when passed new date prop
  useEffect(() => {
    setPickerState((state) => ({
      ...state,
      date: getDate(date),
    }));
  }, [date]);

  const typeWithDefault = type || DEFAULT_TYPE;

  const format = DATE_FORMATS[typeWithDefault];

  const minWithDefault = min
    ? DateTime.fromISO(min)
    : now().minus(DEFAULT_MIN_DATE);

  const maxWithDefault = max
    ? DateTime.fromISO(max)
    : now();

  const range = Interval.fromDateTimes(minWithDefault.startOf('day'), maxWithDefault.endOf('day'));

  const handleOnChange = (values) => {
    const newDate = { ...pickerState.date, ...values };

    setPickerState((state) => ({
      ...state,
      date: newDate,
    }));

    if (isEmpty(type)(newDate)) {
      onChange('');
      return;
    }

    if (isComplete(type)(newDate)) {
      const dateString = DateTime.fromObject(newDate).toFormat(format);
      onChange(dateString);
    }
  };

  const context = {
    onChange: handleOnChange,
    range,
    type,
    ...pickerState,
  };

  return (
    <DatePickerContext.Provider value={context}>
      {children}
    </DatePickerContext.Provider>
  );
};

DatePicker.defaultProps = {
  children: null,
  type: DEFAULT_TYPE,
  date: null,
  min: null,
  max: null,
  onChange: () => {},
};

DatePicker.propTypes = {
  children: PropTypes.node,
  date: PropTypes.string,
  type: PropTypes.oneOf([...Object.keys(DATE_FORMATS)]),
  min: PropTypes.string,
  max: PropTypes.string,
  onChange: PropTypes.func,
};

export default DatePicker;
