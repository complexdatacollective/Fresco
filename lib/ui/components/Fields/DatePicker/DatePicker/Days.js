import { useContext } from 'react';
import { range } from 'lodash';
import { DateTime } from 'luxon';
import DatePickerContext from './DatePickerContext';
import { formatRangeItem } from './helpers';

/**
 * Supplies `days` range based on currently selected month.
 */
const Days = ({ children }) => {
  const { date, range: dateRange } = useContext(DatePickerContext);

  const days = range(1, DateTime.fromObject(date).daysInMonth + 1)
    .map((day) => {
      const d = DateTime.fromObject({ ...date, day });
      if (dateRange.contains(d)) {
        return formatRangeItem(day);
      }
      return formatRangeItem(day, { isOutOfRange: true });
    });

  return children({ days });
};

export default Days;
