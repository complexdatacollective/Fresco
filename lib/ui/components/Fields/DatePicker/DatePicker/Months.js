import { useContext } from 'react';
import { range } from 'lodash';
import { Interval } from 'luxon';
import DatePickerContext from './DatePickerContext';
import { formatRangeItem, getMonthName } from './helpers';

/**
 * Supplies `months` range.
 */
const Months = ({ children }) => {
  const { date, range: dateRange } = useContext(DatePickerContext);

  const months = range(1, 13)
    .map((month) => {
      // Create a month long period
      const m = Interval.after({ ...date, month, day: 1 }, { months: 1 });
      const label = getMonthName(month);
      // if it overlaps min/max period, then this month is valid
      if (!dateRange.overlaps(m) || !date.year) {
        return formatRangeItem(month, { label, isOutOfRange: true });
      }
      return formatRangeItem(month, { label });
    });

  return children({ months });
};

export default Months;
