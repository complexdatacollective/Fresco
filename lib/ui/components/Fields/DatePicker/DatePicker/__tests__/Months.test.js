/* eslint-env jest */

import React from 'react';
import { mount } from 'enzyme';
import DatePicker from '../DatePicker';
import Months from '../Months';

const getSubject = (props) => {
  const mockFunctionalComponent = jest.fn(() => null);

  mount((
    // eslint-disable-next-line react/jsx-props-no-spreading
    <DatePicker {...props}>
      <Months>{mockFunctionalComponent}</Months>
    </DatePicker>
  ));

  return mockFunctionalComponent;
};

describe('<Months>', () => {
  it('provides months', () => {
    const subject = getSubject();

    expect(subject.mock.calls[0][0])
      .toMatchSnapshot();
  });

  it('for empty date, marks all months as out of range', () => {
    const shortMonth = getSubject({ min: '2019-05-01', max: '2019-08-01' });

    expect(shortMonth.mock.calls[0][0])
      .toMatchSnapshot();
  });

  it('marks months as out of range', () => {
    const shortMonth = getSubject({ date: '2019-06-01', min: '2019-05-01', max: '2019-08-01' });

    expect(shortMonth.mock.calls[0][0])
      .toMatchSnapshot();
  });
});
