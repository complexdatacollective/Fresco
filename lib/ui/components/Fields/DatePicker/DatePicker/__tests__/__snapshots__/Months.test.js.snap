// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Months> for empty date, marks all months as out of range 1`] = `
Object {
  "months": Array [
    Object {
      "isOutOfRange": true,
      "label": "January",
      "value": 1,
    },
    Object {
      "isOutOfRange": true,
      "label": "February",
      "value": 2,
    },
    Object {
      "isOutOfRange": true,
      "label": "March",
      "value": 3,
    },
    Object {
      "isOutOfRange": true,
      "label": "April",
      "value": 4,
    },
    Object {
      "isOutOfRange": true,
      "label": "May",
      "value": 5,
    },
    Object {
      "isOutOfRange": true,
      "label": "June",
      "value": 6,
    },
    Object {
      "isOutOfRange": true,
      "label": "July",
      "value": 7,
    },
    Object {
      "isOutOfRange": true,
      "label": "August",
      "value": 8,
    },
    Object {
      "isOutOfRange": true,
      "label": "September",
      "value": 9,
    },
    Object {
      "isOutOfRange": true,
      "label": "October",
      "value": 10,
    },
    Object {
      "isOutOfRange": true,
      "label": "November",
      "value": 11,
    },
    Object {
      "isOutOfRange": true,
      "label": "December",
      "value": 12,
    },
  ],
}
`;

exports[`<Months> marks months as out of range 1`] = `
Object {
  "months": Array [
    Object {
      "isOutOfRange": true,
      "label": "January",
      "value": 1,
    },
    Object {
      "isOutOfRange": true,
      "label": "February",
      "value": 2,
    },
    Object {
      "isOutOfRange": true,
      "label": "March",
      "value": 3,
    },
    Object {
      "isOutOfRange": true,
      "label": "April",
      "value": 4,
    },
    Object {
      "label": "May",
      "value": 5,
    },
    Object {
      "label": "June",
      "value": 6,
    },
    Object {
      "label": "July",
      "value": 7,
    },
    Object {
      "label": "August",
      "value": 8,
    },
    Object {
      "isOutOfRange": true,
      "label": "September",
      "value": 9,
    },
    Object {
      "isOutOfRange": true,
      "label": "October",
      "value": 10,
    },
    Object {
      "isOutOfRange": true,
      "label": "November",
      "value": 11,
    },
    Object {
      "isOutOfRange": true,
      "label": "December",
      "value": 12,
    },
  ],
}
`;

exports[`<Months> provides months 1`] = `
Object {
  "months": Array [
    Object {
      "isOutOfRange": true,
      "label": "January",
      "value": 1,
    },
    Object {
      "isOutOfRange": true,
      "label": "February",
      "value": 2,
    },
    Object {
      "isOutOfRange": true,
      "label": "March",
      "value": 3,
    },
    Object {
      "isOutOfRange": true,
      "label": "April",
      "value": 4,
    },
    Object {
      "isOutOfRange": true,
      "label": "May",
      "value": 5,
    },
    Object {
      "isOutOfRange": true,
      "label": "June",
      "value": 6,
    },
    Object {
      "isOutOfRange": true,
      "label": "July",
      "value": 7,
    },
    Object {
      "isOutOfRange": true,
      "label": "August",
      "value": 8,
    },
    Object {
      "isOutOfRange": true,
      "label": "September",
      "value": 9,
    },
    Object {
      "isOutOfRange": true,
      "label": "October",
      "value": 10,
    },
    Object {
      "isOutOfRange": true,
      "label": "November",
      "value": 11,
    },
    Object {
      "isOutOfRange": true,
      "label": "December",
      "value": 12,
    },
  ],
}
`;
